spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -393,13 +397,67 @@ def sample_allocation(obj, _unused = 0)
           samples.select { |s| s.values[:'heap-live-samples'] == 0 }
         end
 
-        it 'include the stack and sample counts for the objects still left alive' do
-          # We sample from 2 distinct locations
-          expect(heap_samples.size).to eq(2)
+        matcher :approximate_age_based_on_allocation_order do |allocation_order|
+          # To prevent flakiness if other GCs other than the requested ones occur
+          # allow some margin in the matching. We do not expect this test to be
+          # particularly memory heavy so at most we expect one unexpected GC.
+          flaky_margin = 1
+
+          match do |gc_gen_age|
+            expected_age = num_allocations - allocation_order
+            gc_gen_age.between?(expected_age, expected_age + flaky_margin)
+          end
+          description do
+            expected_age = num_allocations - allocation_order
+            "be between #{expected_age} and #{expected_age + flaky_margin}"
+          end
+        end
```

Comment:
So... I'm a bit divided about this. It does work, and it's quite fancy, but I wonder if we could test the same property without needing this kind of machinery.

For instance, if we're testing the age of objects gets calculated correctly, perhaps we could assert in a separate testcase that the ages of the 3 relevant samples are in order of age (e.g. `ages.size == ages.size.uniq && ages.sort == ages`) and that the last age is within the expected value -- something like that?

Reply:
Yeah this is a good suggestion 👍 Done and moved the age accuracy checks into their own example as I felt the existing one was becoming quite bloated.